src/v/storage/disk_log_impl.cc

```diff
@@ -2313,4 +2314,201 @@ void disk_log_impl::set_cloud_gc_offset(model::offset offset) {
+     * retention setting, but we are going to exempt them from forced reclaim
+     * until this bug is fixed to avoid any complications.
+     *
+     * https://github.com/redpanda-data/redpanda/issues/11936
+     */
+    if (config().is_read_replica_mode_enabled()) {
+        vlog(
+          stlog.debug,
+          "Reporting no reclaimable offsets for read replica partition {}",
+          config().ntp());
+        co_return res;
+    }
+
+    /*
+     * calculate the effective local retention. this forces the local retention
+     * override in contrast to housekeeping GC where the overrides are applied
+     * only when local retention is non-advisory.
+     */
+    cfg = apply_base_overrides(cfg);
+    cfg = override_retention_config(cfg);
+    const auto local_retention_offset
+      = co_await maybe_adjusted_retention_offset(cfg);
+
+    /*
+     * when local retention is based off an explicit override, then we treat it
+     * as the partition having a retention hint and use it to deprioritize
+     * selection of data to reclaim over data without any hints.
+     */
+    const auto hinted = has_local_retention_override();
+
+    /*
+     * for a cloud-backed topic the max collecible offset is the threshold below
+     * which data has been uploaded and can safely be removed from local disk.
+     */
+    const auto max_collectible = stm_manager()->max_collectible_offset();
+
+    /*
+     * lightweight segment set copy for safe iteration
+     */
+    fragmented_vector<segment_set::type> segments;
+    for (const auto& seg : _segs) {
+        segments.push_back(seg);
+    }
+
+    /*
+     * currently we use two segments as the low watermark size
+     */
+    std::optional<model::offset> low_watermark_offset;
+    if (segments.size() >= 2) {
+        low_watermark_offset
+          = segments[segments.size() - 2]->offsets().base_offset;
+    }
+
+    vlog(
+      stlog.debug,
+      "Categorizing {} {} segments for {} with local retention {} low "
+      "watermark {} max collectible {}",
+      segments.size(),
+      (hinted ? "hinted" : "non-hinted"),
+      config().ntp(),
+      local_retention_offset,
+      low_watermark_offset,
+      max_collectible);
+
+    /*
+     * categorize each segment.
+     */
+    for (const auto& seg : segments) {
+        const auto usage = co_await seg->persistent_size();
+        const auto seg_size = usage.total();
+
+        /*
+         * the active segment designation takes precedence because it requires
+         * special consideration related to the implications of force rolling.
+         */
+        if (seg == segments.back()) {
+            /*
+             * since the active segment receives all new data at any given time
+             * it may not be fully uploaded to cloud storage so it is hard to
+             * say anything definitive about it. instead, we report its size as
+             * a potential:
+             *
```

Comment:
Upload could potentially lag behind and rolling `segments.back()` won't do much. If this is the case it doesn't make much sense to roll the segment.

Reply:
I believe this is handled below where we check if max collectible has _at least_ made it up to segments.back() / the-active-segment. The condition would then be that _if_ the active segment were rolled then we'd have a point-in-time estimate about how much data would be available once max collectible progressed past the dirty offset.

in any case, this PR does not force roll the active segment. that's a policy enhancement left for later.

does that answer your question / concern?